File: rust/everybody-codes/src/bin/2025/y25d17.rs
use std::collections::BinaryHeap;
use std::hash::Hash;

use everybody_codes::utilities::inputs::get_event_inputs;
use rustc_hash::FxHashSet;
use utilities::structs::grid::{Grid2, GridAdjacent, GridIterator};
use utilities::{
    enums::cardinals::Cardinal,
    structs::{
        coord::Coord2U,
        stopwatch::{ReportDuration, Stopwatch},
        str_grid::StrGrid,
    },
};

type Input<'a> = &'a str;
type Pos = Coord2U;

fn main() {
    let mut stopwatch = Stopwatch::new();
    stopwatch.start();
    let (input1, input2, input3) = get_event_inputs(25, 17);
    println!("Input parsed ({})", stopwatch.lap().report());
    println!("1. {} ({})", part1(&input1), stopwatch.lap().report());
    println!("2. {} ({})", part2(&input2), stopwatch.lap().report());
    println!("3. {} ({})", part3(&input3), stopwatch.lap().report());
    println!("Total: {}", stopwatch.stop().report());
}

trait InRange {
    fn in_range(&self, other: &Self, radius: usize) -> bool;
}

impl InRange for Pos {
    fn in_range(&self, other: &Self, radius: usize) -> bool {
        let x = self.x().abs_diff(other.x());
        let y = self.y().abs_diff(other.y());
        let p = x * x + y * y;
        p <= radius * radius
    }
}

fn part1(notes: Input) -> usize {
    let volcano: StrGrid = notes.into();
    let center = volcano.s.iter().position(|&b| b == b'@').unwrap();
    let center = Pos::from_index(center, volcano.width).unwrap();
    volcano
        .s
        .iter()
        .enumerate()
        .filter(|&(idx, &b)| {
            b != b'@'
                && Pos::from_index(idx, volcano.width)
                    .unwrap()
                    .in_range(&center, 10)
        })
        .map(|(_, &b)| (b - b'0') as usize)
        .sum()
}

fn part2(notes: Input) -> usize {
    let volcano: StrGrid = notes.into();
    let center = volcano.s.iter().position(|&b| b == b'@').unwrap();
    let center = Pos::from_index(center, volcano.width).unwrap();
    let r2: Vec<_> = (0..center.x()).map(|it| it * it).collect();
    let mut destruction = vec![0usize; r2.len()];
    for (idx, &b) in volcano.s.iter().enumerate() {
        if !b.is_ascii_digit() {
            continue;
        }
        let pos = Pos::from_index(idx, volcano.width).unwrap();
        let x = center.x().abs_diff(pos.x());
        let y = center.y().abs_diff(pos.y());
        let p = x * x + y * y;
        let d = (b - b'0') as usize;
        let r = r2.binary_search(&p).unwrap_or_else(|r| r);
        if r < destruction.len() {
            destruction[r] += d;
        }
    }

    destruction
        .iter()
        .enumerate()
        .max_by_key(|&(_, &destruction)| destruction)
        .map(|(radius, &destruction)| radius * destruction)
        .unwrap()
}

fn part3(notes: Input) -> usize {
    let volcano: StrGrid = notes.into();
    let center = volcano.s.iter().position(|&b| b == b'@').unwrap();
    let center = Pos::from_index(center, volcano.width).unwrap();
    let start = volcano.s.iter().position(|&b| b == b'S').unwrap();
    let start = Pos::from_index(start, volcano.width).unwrap();
    let r2: Vec<_> = (0..center.x()).map(|it| it * it).collect();
    let volcano = notes
        .as_bytes()
        .iter()
        .enumerate()
        .filter(|&(_, &b)| b != b'\n')
        .map(|(idx, &b)| {
            let pos = Pos::from_index(idx, volcano.width).unwrap();
            let x = center.x().abs_diff(pos.x());
            let y = center.y().abs_diff(pos.y());
            let p = x * x + y * y;
            let seconds = if b == b'S' {
                0
            } else {
                (b - b'0') as usize
            };
            let r = r2.binary_search(&p).unwrap_or_else(|r| r);
            (r, seconds)
        })
        .try_collect_grid(volcano.width - 1)
        .unwrap();

    let mut min_seconds = 0;

    for radius in 10..(volcano.width() - 1) / 2 - 1 {
        if min_seconds <= 30 * (radius + 1) - 1 {
            match a_star(&volcano, start, center, radius) {
                Ok(seconds) => {
                    return seconds * radius
                },
                Err(seconds) => { min_seconds = seconds; },
            }
        }
    }
    unreachable!()
}

fn a_star(
    volcano: &Grid2<(usize, usize)>,
    start: Pos,
    center: Pos,
    radius: usize,
) -> Result<usize, usize> {
    let heuristic = |pos: Pos, phase: Cardinal| {
        let mut pos = pos;
        let mut h = 0;
        if phase == Cardinal::East {
            let target = center.move_direction(Cardinal::East, radius + 1).unwrap();
            h += pos.manhattan_distance(target);
            pos = target;
        }
        if phase == Cardinal::East || phase == Cardinal::South {
            let target = center.move_direction(Cardinal::South, radius + 1).unwrap();
            h += pos.manhattan_distance(target);
            pos = target;
        }
        if phase != Cardinal::North {
            let target = center.move_direction(Cardinal::West, radius + 1).unwrap();
            h += pos.manhattan_distance(target);
            pos = target;
        }
        h + pos.manhattan_distance(start)
    };

    let mut open: BinaryHeap<State> = BinaryHeap::new();
    let initial_state = State {
        pos: start,
        seconds: 0,
        phase: Cardinal::East,
        f: heuristic(start, Cardinal::East),
    };
    open.push(initial_state);
    let mut closed: FxHashSet<(Pos, Cardinal)> = FxHashSet::default();

    while let Some(State {
        pos,
        seconds,
        phase,
        f: _,
    }) = open.pop()
    {
        if !closed.insert((pos, phase)) {
            continue;
        }

        if pos == start && phase == Cardinal::North {
            return if seconds <= 30 * (radius + 1) - 1 {
                Ok(seconds)
            } else {
                Err(seconds)
            };
        }
        for GridAdjacent {
            index: _,
            pos: adj_pos,
            dir: _,
            value: &(adj_rad, adj_sec),
        } in volcano.adjacent(pos, false).unwrap()
        {
            // abort case 1: ventured into burn area
            if adj_rad <= radius {
                continue;
            }

            // ac2: no significant backtracking once quadrant checkpoint reached
            if match phase {
                Cardinal::North => adj_pos.y() > center.y() + 10,
                Cardinal::East => adj_pos.x() < center.x() - 10,
                Cardinal::South => adj_pos.y() < center.y() - 10,
                Cardinal::West => adj_pos.x() > center.x() + 10,
            } {
                continue;
            }

            let adj_phase = match phase {
                Cardinal::East => if adj_pos.y() == center.y() {
                    Cardinal::South
                } else {
                    phase
                },
                Cardinal::South => {
                    if adj_pos.x() == center.x() {
                        Cardinal::West
                    } else {
                        phase
                    }
                },
                Cardinal::West => {
                    if adj_pos.y() == center.y() {
                        Cardinal::North
                    } else {
                        phase
                    }
                },
                Cardinal::North => phase,
            };

            // ac3: already visited
            if closed.contains(&(adj_pos, adj_phase)) {
                continue;
            }
            let adj_sec = seconds + adj_sec;
            let adj_f = adj_sec + heuristic(adj_pos, adj_phase);

            let adj_state = State {
                pos: adj_pos,
                seconds: adj_sec,
                phase: adj_phase,
                f: adj_f,
            };
            open.push(adj_state);
        }
    }
    Err(usize::MAX)
}

#[derive(Debug, Eq, PartialEq, Hash)]
struct State {
    pos: Pos,
    seconds: usize,
    phase: Cardinal,
    f: usize,
}

impl Ord for State {
    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
        other.f.cmp(&self.f).then(other.seconds.cmp(&self.seconds))
    }
}

impl PartialOrd for State {
    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
        Some(self.cmp(other))
    }
}

#[test]
fn default() {
    let (input1, input2, input3) = get_event_inputs(25, 17);
    assert_eq!(1584, part1(&input1));
    assert_eq!(66183, part2(&input2));
    assert_eq!(42069, part3(&input3));
}

// Input parsed (40μs)
// 1. 1584 (25μs)
// 2. 66183 (74μs)
// 3. 42069 (33.728ms)
// Total: 33.872ms

File: kotlin/everybody_codes/src/main/kotlin/org/gristle/pdxpuzzles/everybodycodes/y2025/Y25D17.kt
package org.gristle.pdxpuzzles.everybodycodes.y2025

import org.gristle.pdxpuzzles.everybodycodes.utilities.Day
import org.gristle.pdxpuzzles.utilities.algorithms.Graph
import org.gristle.pdxpuzzles.utilities.algorithms.Graph.steps
import org.gristle.pdxpuzzles.utilities.enums.Nsew
import org.gristle.pdxpuzzles.utilities.objects.Coord
import org.gristle.pdxpuzzles.utilities.objects.Grid
import org.gristle.pdxpuzzles.utilities.objects.StringGrid
import org.gristle.pdxpuzzles.utilities.objects.toGrid
import kotlin.math.abs

object Y25D17 : Day {
    private fun Coord.inRange(other: Coord, radius: Int): Boolean {
        val dX = abs(x - other.x)
        val dY = abs(y - other.y)
        val p = dX * dX + dY * dY
        return p <= radius * radius
    }

    override fun part1(input: String): Int {
        val volcano = StringGrid(input)
        val centerIdx = volcano.string.indexOf('@')
        val center = Coord.fromIndex(centerIdx, volcano.width + 1)
        return volcano.string
            .filterIndexed { idx, ch ->
                ch != '@' && Coord.fromIndex(idx, volcano.width + 1).inRange(center, 10)
            }.sumOf { ch -> ch - '0' }
    }

    override fun part2(input: String): Int {
        val volcano = StringGrid(input)
        val centerIdx = volcano.string.indexOf('@')
        val center = Coord.fromIndex(centerIdx, volcano.width + 1)
        val r2 = (0 until center.x).map { it * it }
        val destruction = MutableList(r2.size) { 0 }
        for ((idx, ch) in volcano.string.withIndex()) {
            if (!ch.isDigit()) continue
            val pos = Coord.fromIndex(idx, volcano.width + 1)
            val dX = abs(center.x - pos.x)
            val dY = abs(center.y - pos.y)
            val p = dX * dX + dY * dY
            val d = ch - '0'
            val r = r2.binarySearch(element = p)
                .let {
                    if (it >= 0) {
                        it
                    } else {
                        abs(it + 1)
                    }
                }
            if (r < destruction.size) destruction[r] += d
        }
        return destruction.withIndex()
            .maxBy { (_, dest) -> dest }
            .let { (radius, dest) -> radius * dest }
    }

    override fun part3(input: String): Int {
        val sVolcano = StringGrid(input)
        val centerIdx = sVolcano.string.indexOf('@')
        val center = Coord.fromIndex(centerIdx, sVolcano.width + 1)
        val startIdx = sVolcano.string.indexOf('S')
        val start = Coord.fromIndex(startIdx, sVolcano.width + 1)
        val r2 = (0 until center.x).map { it * it }
        val volcano = input
            .mapIndexedNotNull { idx, ch ->
                if (ch == '\n') {
                    null
                } else {
                    val pos = Coord.fromIndex(idx, sVolcano.width + 1)
                    val dX = abs(center.x - pos.x)
                    val dY = abs(center.y - pos.y)
                    val p = dX * dX + dY * dY
                    val seconds = if (ch == 'S') 0 else ch - '0'
                    val r = r2.binarySearch(element = p)
                        .let {
                            if (it >= 0) {
                                it
                            } else {
                                abs(it + 1)
                            }
                        }
                    r to seconds
                }
            }.toGrid(sVolcano.width)

        var minSeconds = 0

        for (radius in 10 until (volcano.width - 1) / 2 - 1) {
            if (minSeconds <= 30 * (radius + 1) - 1) {
                val (success, seconds) = aStar(volcano, start, center, radius)
                if (success) {
                    return seconds * radius
                } else {
                    minSeconds = seconds
                }
            }
        }
        error("Unreachable")
    }

    private fun aStar(
        volcano: Grid<Pair<Int, Int>>,
        start: Coord,
        center: Coord,
        radius: Int
    ): Pair<Boolean, Int> {
        val heuristic: (Pair<Coord, Nsew>) -> Double = { (pos, phase) ->
            var pos = pos
            var h = 0
            if (phase == Nsew.EAST) {
                val target = center.move(phase, radius + 1)
                h += pos.manhattanDistance(target)
                pos = target
            }
            if (phase == Nsew.EAST || phase == Nsew.SOUTH) {
                val target = center.move(Nsew.SOUTH, radius + 1)
                h += pos.manhattanDistance(target)
                pos = target
            }
            if (phase != Nsew.NORTH) {
                val target = center.move(Nsew.WEST, radius + 1)
                h += pos.manhattanDistance(target)
                pos = target
            }
            (h + pos.manhattanDistance(start)).toDouble()
        }

        val endCondition: (Pair<Coord, Nsew>) -> Boolean = { (pos, phase) ->
            pos == start && phase == Nsew.NORTH
        }

        val getEdges: (Pair<Coord, Nsew>) -> List<Graph.Edge<Pair<Coord, Nsew>>> = { (pos, phase) ->
            volcano.getNeighborsIndexedValue(pos)
                .mapNotNull { (idx, value) ->
                    val adjPos = Coord.fromIndex(idx, volcano.width)
                    val (adjRad, adjSec) = value

                    // abort case 1
                    if (adjRad <= radius) return@mapNotNull null

                    // ac2
                    val backtracking = when (phase) {
                        Nsew.NORTH -> adjPos.y > center.y + 10
                        Nsew.SOUTH -> adjPos.y < center.y - 10
                        Nsew.EAST -> adjPos.x < center.x - 10
                        Nsew.WEST -> adjPos.x > center.x + 10
                    }
                    if (backtracking) return@mapNotNull null

                    val adjPhase = when (phase) {
                        Nsew.NORTH -> phase
                        Nsew.SOUTH -> if (adjPos.x == center.x) {
                            Nsew.WEST
                        } else {
                            phase
                        }
                        Nsew.EAST -> if (adjPos.y == center.y) {
                            Nsew.SOUTH
                        } else {
                            phase
                        }
                        Nsew.WEST -> if (adjPos.y == center.y) {
                            Nsew.NORTH
                        } else {
                            phase
                        }
                    }

                    Graph.Edge(adjPos to adjPhase, adjSec.toDouble())
                }
        }

        val seconds = Graph
            .aStar(
                start to Nsew.EAST,
                heuristic,
                endCondition,
                defaultEdges = getEdges,
            ).steps()
        val success = seconds <= 30 * (radius + 1) - 1
        return success to seconds
    }
}

fun main() = Day.runDay(Y25D17::class)

//    Quest 1: 1584 (1ms)
//    Quest 2: 66183 (7ms)
//    Quest 3: 42069 (171ms)
//    Total time: 181ms